include/swift/Runtime/Concurrent.h

```diff
@@ -406,6 +409,102 @@ class ConcurrentMap
+/// writes.
+template <class ElemTy> struct ConcurrentReadableArray {
+private:
+  /// The struct used for the array's storage. The `Elem` member is
+  /// considered to be the first element of a variable-length array,
+  /// whose size is determined by the allocation. The `Capacity` member
+  /// from `ConcurrentReadableArray` indicates how large it can be.
+  struct Storage {
+    std::atomic<size_t> Count;
+    typename std::aligned_storage<sizeof(ElemTy), alignof(ElemTy)>::type Elem;
+    
+    ElemTy *data() {
+      return reinterpret_cast<ElemTy *>(&Elem);
+    }
+  };
+  
+  size_t Capacity;
+  std::atomic<size_t> ReaderCount;
+  std::atomic<Storage *> Elements;
+  Mutex WriterLock;
+  std::vector<Storage *> FreeList;
+  
+  Storage *allocate(size_t capacity) {
+    auto size = sizeof(Storage) + (capacity - 1) * sizeof(Storage().Elem);
+    auto *ptr = reinterpret_cast<Storage *>(malloc(size));
+    if (!ptr) swift::crash("Could not allocate memory.");
+    ptr->Count.store(0, std::memory_order_relaxed);
+    return ptr;
+  }
+  
+  void deallocate(Storage *storage) {
+    if (storage == nullptr) return;
+    
+    auto *data = storage->data();
+    for (size_t i = 0; i < storage->Count; i++) {
+      data[i].~ElemTy();
+    }
+    free(storage);
+  }
+  
+public:
+  void push_back(const ElemTy &elem) {
+    ScopedLock guard(WriterLock);
+    
+    auto *storage = Elements.load(std::memory_order_relaxed);
+    auto count = storage ? storage->Count.load(std::memory_order_relaxed) : 0;
+    if (count >= Capacity) {
+      auto newCapacity = std::max((size_t)16, count * 2);
+      auto *newStorage = allocate(newCapacity);
+      if (storage) {
+        std::copy(storage->data(), storage->data() + count, newStorage->data());
+        FreeList.push_back(storage);
+      }
+      
+      storage = newStorage;
+      Capacity = newCapacity;
+      Elements.store(storage, std::memory_order_release);
+    }
+    
+    auto Count = storage->Count.load(std::memory_order_relaxed);
+    new(&storage->data()[Count]) ElemTy(elem);
+    storage->Count.store(Count + 1, std::memory_order_release);
+    
+    if (ReaderCount.load(std::memory_order_relaxed) == 0)
+      for (Storage *storage : FreeList)
+        deallocate(storage);
+  }
+  
+  /// Read the contents of the array. The parameter `f` is called with
+  /// two parameters: a pointer to the elements in the array, and the
+  /// count. This represents a snapshot of the contents at the time
+  /// `read` was called. The pointer becomes invalid after `f` returns.
+  template <class F> auto read(F f) -> decltype(f(nullptr, 0)) {
+    ReaderCount.fetch_add(1, std::memory_order_relaxed);
+    auto *storage = Elements.load(std::memory_order_consume);
+    auto count = storage->Count.load(std::memory_order_acquire);
+    auto *ptr = storage->data();
+    
+    decltype(f(nullptr, 0)) result = f(ptr, count);
+    
+    ReaderCount.fetch_sub(1, std::memory_order_relaxed);
```

Comment:
My impression is that this should be `release` (and the `load` in `push_back` above an `acquire`), so that there's a barrier for `f`s reads and thus ensure that they have finished before calling `deallocate`? Or am I missing something?

Hm, now that I think about it, the `fetch_add` on `ReaderCount` above should probably also be `acquire`, because I don't think there's anything stopping that being moved all the way past the `Count` load and past the `f` call. (I may be misremembering something special about atomic accesses before `acquire` ones, though.)